build/docs/deploy.adoc

```diff
@@ -0,0 +1,36 @@
+Deploy Infrastructure with Terraform.
+
+This tasks supports applying a Terraform configuration.
+
+NOTICE: This is experimental and anything is subject to change or even removal. 
+
+The Terraform configuration is associated with a state file unique to the
+target-environment. 
+
+This tasks provides a terraform kubernetes backend (see https://developer.hashicorp.com/terraform/language/settings/backends/kubernetes). The `secret_suffix` used is `component`-`target-environment`.
+In the future other backends shall be supported as needed where S3 support is an obvious candidate
+
+This task runs the following terraform commands in sequence:
+
+- `terraform init` with parameters to configure the backend and with env variable TF_PLUGIN_CACHE_DIR set to cache the provider plugins. 
+
+- `terraform plan` if no changes are detected or parameter `plan-only` is true tno actual deployments happen.
+
+- `terraform apply` to apply the changes to the target environment.
+
+It is assumed that secrets needed to connected to the infrastructure managed by terraform are provided with environment variables. The task by default expects a kubernetes secret which is used to derived the needed environment variables from. This can be switched off by setting the `env-from-secret` to "false" in case variables are already provided by other means (such as a podTemplate) or not needed.
+
+This mechanism is also the means to provide terraform input variables.
+
+If the pipeline runs for a repository defining subrepos in its `ods.y(a)ml`
+file, then any terraform configs in those subrepos are processed as well. Note that parameters definitions considered are only the ones defined in the repository for which the pipeline
+runs. Therefore, if you use an umbrella repository to promote an
+application consisting of multiple repositories, the umbrella repository
+needs to define the environment specific values for the subcomponents.
```

Comment:
Could this lead to conflicting env names?

Reply:
This would be possible, for example if one had multiple AWS accounts. On the other hand perhaps one can then survive by having variable names which encode this in the variable names. Alternatively one could have the component name to become part of the secret. We could make this optional perhaps by changing the `envFromSecret` boolean flag to be `envFromSecretPrefix` string which might default to 'ods-pipeline-terraform-envs' and the final secret name adds the `-{{target-environment}}` to it. If envFromSecretPrefix is empty one could disable using secrets for env. Not sure whether this should be added from the get go.

Reply:
I guess we can leave like it is for now but I am pretty sure this will become an issue. I wasn't thinking about multiple AWS accounts but rather two components using the same env var key.

Reply:
If my proposal makes sense I could update this PR. If it needs more work I'd rather handle it in a different issue.
I was under the impression that one can have environments (with secrets) defined on an organization level but now found https://github.com/orgs/community/discussions/15379

Reply:
I lean towards doing it later, and starting with what you have.